src/hotspot/share/code/codeCache.cpp

```diff
@@ -1430,27 +1430,66 @@ void CodeCache::print() {
 #ifndef PRODUCT
   if (!Verbose) return;
 
-  CodeBlob_sizes live;
-  CodeBlob_sizes dead;
+  static_assert(0 < CompLevel_simple && CompLevel_simple <= CompLevel_full_optimization, "tier range check");
+  CodeBlob_sizes live[CompLevel_full_optimization];
+  CodeBlob_sizes dead[CompLevel_full_optimization];
+  CodeBlob_sizes runtimeStub;
+  CodeBlob_sizes uncommonTrapStub;
+  CodeBlob_sizes deoptimizationStub;
+  CodeBlob_sizes adapter;
+  CodeBlob_sizes bufferBlob;
 
   FOR_ALL_ALLOCABLE_HEAPS(heap) {
     FOR_ALL_BLOBS(cb, *heap) {
-      if (!cb->is_alive()) {
-        dead.add(cb);
-      } else {
-        live.add(cb);
+      if (cb->is_nmethod()) {
+        const int level = cb->as_nmethod_or_null()->comp_level();
+        if (CompLevel_simple <= level && level <= CompLevel_full_optimization) {
+          if (!cb->is_alive()) {
+            dead[level - 1].add(cb);
+          } else {
+            live[level - 1].add(cb);
+          }
+        }
+      } else if (cb->is_runtime_stub()) {
+        runtimeStub.add(cb);
+      } else if (cb->is_deoptimization_stub()) {
+        deoptimizationStub.add(cb);
+      } else if (cb->is_uncommon_trap_stub()) {
+        uncommonTrapStub.add(cb);
+      } else if (cb->is_adapter_blob()) {
+        adapter.add(cb);
+      } else if (cb->is_buffer_blob()) {
+        bufferBlob.add(cb);
       }
     }
   }
 
-  tty->print_cr("CodeCache:");
   tty->print_cr("nmethod dependency checking time %fs", dependentCheckTime.seconds());
-
-  if (!live.is_empty()) {
-    live.print("live");
+  for (int i = CompLevel_simple; i <= CompLevel_full_optimization; i++) {
+    tty->print_cr("Tier %d:", i);
+    if (!live[i - 1].is_empty()) {
```

Comment:
your policy is to skip empty entries. I suggest you move is_empty() logic to CodeBlob_sizes::print(). 
one side, it would print out "#0 live: empty"  instead of skipping it. that's informative. 
secondly, it should simplify your code here.